sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java

```diff
@@ -0,0 +1,336 @@
+    }
+
+    private static QuerySpec mergeQuerySpec(@Nullable QuerySpec querySpec1, QuerySpec querySpec2) {
+        if (querySpec1 == null) {
+            return querySpec2;
+        }
+
+        QuerySpec querySpec = new QuerySpec()
+            .outputs(querySpec1.outputs())
+            .where(mergeWhere(querySpec1, querySpec2))
+            .orderBy(mergeOrderBy(querySpec1, querySpec2))
+            .offset(mergeOffset(querySpec1, querySpec2))
+            .limit(mergeLimit(querySpec1, querySpec2))
+            .groupBy(pushGroupBy(querySpec1, querySpec2))
+            .having(pushHaving(querySpec1, querySpec2))
+            .hasAggregates(querySpec1.hasAggregates() || querySpec2.hasAggregates());
+
+        SymbolNormalizer.normalize(querySpec);
+
+        return querySpec;
+    }
+
+    private static WhereClause mergeWhere(QuerySpec querySpec1, QuerySpec querySpec2) {
+        WhereClause where1 = querySpec1.where();
+        WhereClause where2 = querySpec2.where();
+
+        if (!where1.hasQuery() || where1 == WhereClause.MATCH_ALL) {
+            return where2;
+        } else if (!where2.hasQuery() || where2 == WhereClause.MATCH_ALL) {
+            return where1;
+        }
+
+        return new WhereClause(AndOperator.join(ImmutableList.of(where2.query(), where1.query())));
+    }
+
+    @Nullable
+    private static OrderBy mergeOrderBy(QuerySpec querySpec1, QuerySpec querySpec2) {
+        if (!querySpec1.orderBy().isPresent()) {
+            return querySpec2.orderBy().orNull();
+        } else if (!querySpec2.orderBy().isPresent()) {
+            return querySpec1.orderBy().orNull();
+        }
+
+        OrderBy orderBy1 = querySpec1.orderBy().get();
+        OrderBy orderBy2 = querySpec2.orderBy().get();
+
+        List<Symbol> orderBySymbols = orderBy2.orderBySymbols();
+        List<Boolean> reverseFlags = Booleans.asList(orderBy2.reverseFlags());
+        List<Boolean> nullsFirst = Arrays.asList(orderBy2.nullsFirst());
+
+        for (int i = 0; i < orderBy1.orderBySymbols().size(); i++) {
+            Symbol orderBySymbol = orderBy1.orderBySymbols().get(i);
+            int idx = orderBySymbols.indexOf(orderBySymbol);
+            if (idx == -1) {
+                orderBySymbols.add(orderBySymbol);
+                reverseFlags.add(orderBy1.reverseFlags()[i]);
+                nullsFirst.add(orderBy1.nullsFirst()[i]);
+            } else {
+                if (reverseFlags.get(idx) != orderBy1.reverseFlags()[i]) {
+                    throw new AmbiguousOrderByException(orderBySymbol);
+                }
+                if (nullsFirst.get(idx) != orderBy1.nullsFirst()[i]) {
+                    throw new AmbiguousOrderByException(orderBySymbol);
+                }
+            }
+        }
+
+        return new OrderBy(orderBySymbols, Booleans.toArray(reverseFlags), nullsFirst.toArray(new Boolean[0]));
+    }
+
+    @Nullable
+    private static Integer mergeOffset(QuerySpec querySpec1, QuerySpec querySpec2) {
```

Comment:
offset is a primitive int, so this method could also return a primitive int and isn't nullable